### PR TITLE
Convert Region and Graphics to use CsWin32

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Converts Region and Graphics to mostly use CsWin32 (modulo enumerating metafiles).

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/10628)

### CODE REVIEW

**Comment by JeremyKuhne** [comment]
VS hangs trying to do the diff on Graphics. 😆 Note that I've done some GC.KeepAlive handling in error helper methods to reduce the number of places I have to type it (notably for the `this` pointer). Most methods are only called once so I'm generally not going to bother creating a second wrapper to keep objects from being collected.

File: src/System.Drawing.Common/src/misc/DpiHelper.cs [hunk 1]
```diff
@@ -25,7 +22,7 @@ internal static class DpiHelper
 
     private static double s_logicalToDeviceUnitsScalingFactorX;
     private static double s_logicalToDeviceUnitsScalingFactorY;
-    private static InterpolationMode s_interpolationMode = InterpolationMode.Invalid;
+    private static Drawing2D.InterpolationMode s_interpolationMode = Drawing2D.InterpolationMode.Invalid;
```
Comment (Tanya-Solyanik) [inline src/System.Drawing.Common/src/misc/DpiHelper.cs:25]: Why didn't you add using System.Drawing.Drawing2D?

File: src/System.Drawing.Common/src/Interop/Windows/Gdi32/Interop.OffsetViewportOrgEx.cs [hunk 2]
```diff
@@ -19,9 +18,9 @@ internal static partial class Gdi32
             IntPtr hdc,
             int x,
             int y,
-            ref Point lppt);
+            ref System.Drawing.Point lppt);
```
Comment (Tanya-Solyanik) [inline src/System.Drawing.Common/src/Interop/Windows/Gdi32/Interop.OffsetViewportOrgEx.cs:21]: What was the reason for this change?

**Review by Tanya-Solyanik — COMMENTED** [review]
I only took a quick look.

File: src/System.Drawing.Common/src/misc/DpiHelper.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/misc/DpiHelper.cs:25]: This is resolving conflict. It is included, there are two `InterpolationMode` definitions. It just so happens that these are the exact same names Gdi+ uses.

File: src/System.Drawing.Common/src/Interop/Windows/Gdi32/Interop.OffsetViewportOrgEx.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/Interop/Windows/Gdi32/Interop.OffsetViewportOrgEx.cs:21]: Conflict. The native GDI+ type is `Point` and `PointF`. (`Rect` and `RectF` are the rectangles)

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs [hunk 3]
```diff
@@ -359,52 +364,38 @@ public float DpiY
...
 
-            // GDI+ interprets the value of InterpolationMode and sets a value accordingly.
-            // Libgdiplus does not, so do this manually here.
-            switch (value)
-            {
-                case InterpolationMode.Default:
-                case InterpolationMode.Low:
-                    value = InterpolationMode.Bilinear;
-                    break;
-                case InterpolationMode.High:
```
Comment (Tanya-Solyanik) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.cs:390]: This switch is being replaced by a cast, is this correct?

Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs:390]: It was a hack for Unix

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs [hunk 4]
```diff
@@ -897,748 +782,710 @@ public void DrawArc(Pen pen, int x, int y, int width, int height, int startAngle
...
-            CheckErrorStatus(Gdip.GdipFillPolygonI(
-                new HandleRef(this, NativeGraphics),
-                new HandleRef(brush, brush.NativeBrush),
-                p, points.Length,
-                fillMode));
+            CheckErrorStatus(PInvoke.GdipFillPolygonI(
+                NativeGraphics,
+                (GpBrush*)brush.NativeBrush,
+                (GdiPlus.Point*)p, points.Length,
+                (GdiPlus.FillMode)fillMode));
```
Comment (Tanya-Solyanik) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.cs:1318]: GC.KeepAlive(brush)?

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs [hunk 5]
```diff
@@ -897,748 +782,710 @@ public void DrawArc(Pen pen, int x, int y, int width, int height, int startAngle
...
         {
-            CheckErrorStatus(Gdip.GdipFillClosedCurve(
-                new HandleRef(this, NativeGraphics),
-                new HandleRef(brush, brush.NativeBrush),
-                p, points.Length));
+            CheckErrorStatus(PInvoke.GdipFillClosedCurve(
+                NativeGraphics,
+                (GpBrush*)brush.NativeBrush,
+                (GdiPlus.PointF*)p, points.Length));
         }
```
Comment (Tanya-Solyanik) [inline src/System.Drawing.Common/src/System/Drawing/Graphics.cs:1426]: GC.KeepAlive(brush)?

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs:1318]: Good catch, I'll add it in my next change

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs:1426]: Yes, I'll add it in the next change

File: src/System.Drawing.Common/src/System/Drawing/Region.cs [hunk 6]
```diff
@@ -95,241 +105,212 @@ public void Dispose()
     private void Dispose(bool disposing)
     {
 #if FINALIZATION_WATCH
-        Debug.WriteLineIf(!disposing && NativeRegion != IntPtr.Zero, $"""
+        Debug.WriteLineIf(!disposing && NativeRegion is null, $"""
```
Comment (lonitra) [inline src/System.Drawing.Common/src/System/Drawing/Region.cs:108]:
```suggestion
        Debug.WriteLineIf(!disposing && NativeRegion is not null, $"""
```

**Review by lonitra — APPROVED** [review]
Some small comments, but these can be addressed in next change

File: src/System.Drawing.Common/src/System/Drawing/Graphics.cs [hunk 7]
```diff
@@ -897,748 +782,710 @@ public void DrawArc(Pen pen, int x, int y, int width, int height, int startAngle
...
         fixed (Point* p = points)
         {
-            CheckErrorStatus(Gdip.GdipFillClosedCurveI(
-                new HandleRef(this, NativeGraphics),
-                new HandleRef(brush, brush.NativeBrush),
-                p, points.Length));
+            CheckErrorStatus(PInvoke.GdipFillClosedCurveI(
+                NativeGraphics,
+                (GpBrush*)brush.NativeBrush,
+                (GdiPlus.Point*)p, points.Length));
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Graphics.cs:1466]: I'll double check the KeepAlives in the next PR.

File: src/System.Drawing.Common/src/System/Drawing/Region.cs (hunk 6 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Region.cs:108]: Thanks, I'll get it in the next. I have a feeling FINALIZAION_WATCH has other breaks. I'd like to make it more general too. Maybe see if we can add a base class in DEBUG only.

**Review by Tanya-Solyanik — APPROVED** [review]

**Pull request merged**